Update Grand Potential Sintering model with strict mass conservation

d-hermawan opened the pull request:
## Reason
The current GrandPotentialKernelAction sintering configuration for chemical potential evolution may result in mass loss due to discretization error.

## Design
Following the Grand Potential Phase Field formulation, the chemical potential equation is changed such that the relationship between concentration and chemical potential is specified at the expense of additional computational cost.

## Impact
- add additional properties in GrandPotentialSinteringMaterial associated with strict mass conservation
- add switch option in GrandPotentialKernelAction for strict mass conservation option
- add a test file and its gold file for the strict mass conservation option in GrandPotentialKernelAction under GrandPotentialPFM test

closes #23329

d-hermawan committed:
Update Grand Potential Sintering model with strict mass conservation

- add additional properties in GrandPotentialSinteringMaterial associated with strict mass conservation
- add switch option in GrandPotentialKernelAction
- add a test file and its gold exodus file for strict mass conservation option in GrandPotentialKernelAction under GrandPotentialPFM test

closes #23329

d-hermawan committed:
Revert "add issue number to tests"

This reverts commit 9abf124f0273c6eae0a1a76370d30b5629a5efcd.

d-hermawan committed:
Update Grand Potential Sintering model with strict mass conservation

- add additional properties in GrandPotentialSinteringMaterial
- add switch option in GrandPotentialKernelAction and add the strict mass conservation formulation
- add test file and its gold file for the strict mass conservation case, along with issue number in tests

closes #23329

d-hermawan commented:
@laagesen please review

laagesen commented:
Thanks for submitting @d-hermawan - will try to get to it tomorrow but it may be Monday

laagesen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -244,50 +262,136 @@ GrandPotentialKernelAction::act()
+        InputParameters params = _factory.getValidParams("MatReaction");
+        params.set<NonlinearVariableName>("variable") = w_names[i];
+        params.set<std::vector<VariableName>>("v") = v4;
+        params.set<MaterialPropertyName>("mob_name") = mob_cw;
```
Here we just want `mob_name` to be a material property that is `-1` everywhere. Right now it is defined and calculated in `GrandPotentialSinteringMaterial`, but we create a lot of extra variables doing that. I can't find an example that would show that we could just define it as `-1` here, maybe @dschwen might have a suggestion?

dschwen replied:
Just set it to `"-1"`

laagesen commented on modules/phase_field/include/materials/GrandPotentialSinteringMaterial.h:
```diff
@@ -111,6 +111,17 @@ class GrandPotentialSinteringMaterial : public DerivativeMaterialInterface<Mater
+  MaterialProperty<Real> & _matreacts;
+  
+  /// mobility between c and w coupling
+  MaterialProperty<Real> & _mob_conc_w;
```
This is just a material property that is going to be `-1` everywhere, I am hoping we can define it right in the action rather than declaring it and computing here and passing in to the action. Please see further comments below

laagesen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -65,8 +66,20 @@ GrandPotentialKernelAction::validParams()
+  parameters.addParam<std::vector<NonlinearVariableName>>(
+      "concentrations", "List of concentration variables for strict mass conservation");
+  parameters.addParam<std::vector<MaterialPropertyName>>(
+      "body_forces", "List of body forces coefficients for strict mass conservation formulation."
```
Please changes names as described above to `h_c_min` or something along those lines and also the description

laagesen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -65,8 +66,20 @@ GrandPotentialKernelAction::validParams()
+  parameters.addParam<std::vector<MaterialPropertyName>>(
+      "body_forces", "List of body forces coefficients for strict mass conservation formulation."
+                     "Place in same order as switching_function_names.");
+  parameters.addParam<std::vector<MaterialPropertyName>>(
```
Please change names as described above to `h_over_kVa` or something like that

laagesen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -65,8 +66,20 @@ GrandPotentialKernelAction::validParams()
+      "mat_reacts", "List of mat react coefficients for strict mass conservation formulation."
+                     "Place in same order as switching_function_names.");
+  parameters.addParam<MaterialPropertyName>(
+      "mobility_cw", "Lcw", "Name of mobility to be used with concentration-chempot coupling."
```
Hopefully we can remove this an just create a material property for `-1` in the action itself

laagesen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -99,6 +112,11 @@ GrandPotentialKernelAction::act()
   auto displaced_mesh = getParam<bool>("use_displaced_mesh");
   auto aniso = getParam<MultiMooseEnum>("anisotropic");
+  const auto c_names = getParam<std::vector<NonlinearVariableName>>("concentrations");
+  const auto MRj = getParam<std::vector<MaterialPropertyName>>("mat_reacts");
```
Please change name

laagesen commented on modules/phase_field/src/materials/GrandPotentialSinteringMaterial.C:
```diff
@@ -102,6 +102,11 @@ GrandPotentialSinteringMaterial::GrandPotentialSinteringMaterial(const InputPara
     _dkappa(declarePropertyDerivative<Real>("kappa", _phi_name)),
     _d2kappa(declarePropertyDerivative<Real>("kappa", _phi_name, _phi_name)),
     _gamma(declareProperty<Real>("gamma")),
+    _bodyforcev(declareProperty<Real>("BFv")),
```
Please change names

laagesen commented on modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:
```diff
@@ -0,0 +1,230 @@
+      mass_conservation = 'true'
+      concentrations = 'c'
+      body_forces = 'BFv BFs'
+      mat_reacts = 'MRv MRs'
```
Please change names

laagesen commented on modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:
```diff
@@ -0,0 +1,230 @@
+  
+  start_time = 0
+  dtmax = 0.05
+  num_steps = 10
```
2 or 3 time steps should be enough, unless you saw that longer than that was necessary for the non-mass conserved version to show violations of mass conservation

laagesen commented on modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:
```diff
@@ -0,0 +1,230 @@
+  dtmax = 0.05
+  num_steps = 10
+
+  [./TimeStepper]
```
I would remove this and keep a fixed time step, just to keep the test input file as short as possible

laagesen commented on modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:
```diff
@@ -0,0 +1,230 @@
+
+  l_tol = 1e-8
+  nl_rel_tol = 1e-8
+  nl_abs_tol = 1e-8
```
should not need this since you are not running to steady state

laagesen commented on modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:
```diff
@@ -0,0 +1,230 @@
+  [../]
+[]
+
+[Debug]
```
Please remove to keep test input file short

laagesen commented on modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:
```diff
@@ -0,0 +1,230 @@
+  [../]
+[]
+
+[Functions]
```
you can set T = 1600 in the initial conditions for the AuxVariable so you can remove this to keep the input file short

laagesen commented on modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:
```diff
@@ -0,0 +1,230 @@
+  [../]
+  [./T]
+    order = CONSTANT
+    family = MONOMIAL
```
can set `initial_condition = 1600` here

laagesen commented on modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:
```diff
@@ -0,0 +1,230 @@
+    variable = bnds
+    execute_on = 'initial timestep_end'
+  [../]
+  [./T_aux]
```
can remove this by setting T in initial conditions

laagesen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -244,50 +262,136 @@ GrandPotentialKernelAction::act()
+        kernel_name = "Coupled_" + w_names[i] + "_" + all_etas[j];
+        _problem->addKernel("CoupledSwitchingTimeDerivative", kernel_name, params);
+      } //     for (unsigned int j = 0; j < n_etas + n_grs; ++j) coupledSwitchingTimeDeriv
+    } //for (unsigned int i = 0; i < n_w; ++i) chempot
```
It looks like there are some misaligned braces in a few places. This can be fixed by running the Clang-format utility on all your modified files. There are extensions to most text editors, such as VSCode or Atom, that will allow you to run this within the editor. The MOOSE test suite will also check for conformance with clang-format as part of the test suite and throw and error if it does not comply, it will give you the option to apply a patch to fix it, but it's easier just to install a plugin to your text editor and run prior to submitting

dschwen replied:
And please remove the commented out loop headers. Indentation should make this clear enough.

dschwen replied:
While you're at it, please also format the input files using the MOOSE VSCode/Atom plugin (or the `hit format file.i` command.

laagesen reviewed (changes requested):
Please also update the documentation file `GrandPotentialKernelAction.md` to reflect the changes to the action

laagesen commented:
Also please make sure your commit message ("Update Grand Potential Sintering model with strict mass conservation option") references the issue number 22329, so you will want to append #22329 to the end of the commit message. Also please make sure there are no trailing whitespaces at the end. I am not sure if you are currently able to see testing messages  from civet (https://civet.inl.gov/job/1330302/) so I am going to paste them here:

##########################################################################
ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
Merge branch 'dhermawa' of https://github.com/d-hermawan/moose into test
Update Grand Potential Sintering model with strict mass conservation option

ERROR: The following files contain trailing whitespace after applying your patch:
	modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
##########################################################################

dschwen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -15,10 +15,11 @@
 
 registerMooseAction("PhaseFieldApp", GrandPotentialKernelAction, "add_kernel");
 
+template <>
```
This should be a static member. The global function template approach for validParams has long been deprecated.

dschwen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -15,10 +15,11 @@
+template <>
 InputParameters
-GrandPotentialKernelAction::validParams()
+validParams<GrandPotentialKernelAction>()
```
please undo these changes

dschwen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -99,6 +112,11 @@ GrandPotentialKernelAction::act()
   auto implicity = getParam<bool>("implicit");
   auto displaced_mesh = getParam<bool>("use_displaced_mesh");
   auto aniso = getParam<MultiMooseEnum>("anisotropic");
+  const auto c_names = getParam<std::vector<NonlinearVariableName>>("concentrations");
```
I believe this will always at least return an empty vector.

dschwen commented on modules/phase_field/src/action/GrandPotentialKernelAction.C:
```diff
@@ -244,50 +262,136 @@ GrandPotentialKernelAction::act()
+          kernel_name = "MBD_" + w_names[i] + "_" + all_etas[j];
+          _problem->addKernel("MaskedBodyForce", kernel_name, params);
+        }
+      } // for (unsigned int i = 0; i < n_w; ++i)
+    } // case true
+  } // switch
 } // GrandPotentialKernelAction::act()
```
Please remove all these end-block comments. If the code is not readable enough  with just the correct indentation then the blocks span too many lines and the code should be refactored.

d-hermawan committed:
- regold several test files that requires GrandPotentialSinteringMaterial
- clean up conditional in GrandPotentialSinteringMaterial for mass conservation and solid free energy options

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1377354/) on bfacf14 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23331/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23331/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bb9ffc7e91904ae40a0dd5badd8d46ba0e4c381f`

dhermawa commented:
@laagesen please review

laagesen commented:
@dhermawa glad you were able to get the changes pushed and thanks for making the updates! Before we proceed, can you please see the previous comment from @moosebuild about style changes? This is saying that the code still needs to be formatted to match the clang-format standards. You can do this either by applying the patch described in the latest @moosebuild comment, or by using the clang-format capability in VSCode as we talked about in our meeting the week before last.

Also the documentation will need to be updated, I don't see any updates to it in the latest push. Please update the documentation file GrandPotentialKernelAction.md to reflect the changes to the action. This file can be found at `moose/modules/phase_field/doc/content/source/action/GrandPotentialKernelAction.md`

dhermawa commented:
@laagesen I've updated the documentation and apply the clang-format changes, please review.

laagesen reviewed (changes requested):
@dhermawa running the test harness brought out a few issues I didn't notice first time around, sorry about that

dhermawa commented:
@laagesen I've addressed the comments, please review

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1378883/) on d5e1ee2 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23331/site)

This comment will be updated on new commits.

dhermawa commented:
@laagesen please review

dhermawa commented:
@laagesen I have rerun the gold file using "use_legacy_material_output" so it matches with PhaseFieldApp.C in moose base code, please activate the tests again

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1393750/) on caeb1ed wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>880a12</th>
      <th colspan=4>#23331 caeb1e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.79%</td>
      <td>84.79%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>83682</td>
      <td>83681</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15012</td>
      <td>15013</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23331/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23331/coverage/framework)


## Modules coverage

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>880a12</th>
      <th colspan=4>#23331 caeb1e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.58%</td>
      <td>89.61%</td>
      <td>+0.03%</td>
      <td>97.20%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12460</td>
      <td>12532</td>
      <td>+72</td>
      <td>104</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1450</td>
      <td>1453</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23331/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23331/coverage/phase_field)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23331/coverage/xfem)

</details>



This comment will be updated on new commits.

dhermawa commented:
@laagesen I am unsure how to proceed with the docker error

laagesen commented:
OK looks like we are good now! @loganharbour or @dschwen could you please merge?

dschwen commented on modules/phase_field/src/actions/GrandPotentialKernelAction.C:
```diff
@@ -244,50 +261,133 @@ GrandPotentialKernelAction::act()
-    params.set<bool>("use_displaced_mesh") = displaced_mesh;
-    kernel_name = "ChiDt_" + w_names[i];
-    _problem->addKernel("SusceptibilityTimeDerivative", kernel_name, params);
+    std::vector<VariableName> v3;
+    v3.resize(n_w);
+    for (unsigned int i = 0; i < n_w; ++i)
+      v3[i] = w_names[i];
```
```suggestion
```

dhermawa replied:
i tried implementing your suggestion, but it seems that because w_names is vector<NonlinearVariableName>, I cannot use it to set the kernel parameters that requires vector<VariableName>, so the assignment fails to compile.

dschwen commented on modules/phase_field/src/actions/GrandPotentialKernelAction.C:
```diff
@@ -244,50 +261,133 @@ GrandPotentialKernelAction::act()
+      // MatDiffusion concentration (coupled with chempot)
+      params = _factory.getValidParams("MatDiffusion");
+      params.set<NonlinearVariableName>("variable") = c_names[i];
+      params.set<std::vector<VariableName>>("v") = v3;
```
```suggestion
      params.set<std::vector<VariableName>>("v") = w_names;
```

same for `v4`. This is just a suggestion. It seems to me that those vectors are not necessary. FYI, you could also just copy a vactor by assignment: `auto v3 = w_names;` . No need for the loops.

dschwen commented on modules/phase_field/src/materials/GrandPotentialSinteringMaterial.C:
```diff
@@ -113,11 +119,16 @@ GrandPotentialSinteringMaterial::GrandPotentialSinteringMaterial(const InputPara
   if ((_switch > 1.0) || (_switch < 0.0))
     mooseError("GrandPotentialSinteringMaterial: surface_switch_value should be between 0 and 1");
 
+  if ((_mass_conservation) && (_solid_energy > 0))
```
FTI this would be fine, too (and preferred_

```suggestion
  if (_mass_conservation && _solid_energy > 0)
```

dschwen reviewed (dismissed):
Looks fine to me. I leave it up to you if you want to make further changes.

dhermawa commented:
@laagesen i have reverted back and recompile it to make sure it works. please activate the test

dschwen reviewed (approved):
Looks ok now. @laagesen please resolve your conversations of your comments are addressed.

laagesen commented:
Trying to figure out how to mark Outdated conversations as resolved (other than just deleting them)...

loganharbour commented:
Larry doesn't have the ability to resolve. He's looked through and verified that they can be resolved so I'll resolve them

loganharbour commented:
On to you for the magic button @dschwen
